firebase-firestore/src/androidTest/java/com/google/firebase/firestore/QueryTest.java

```diff
@@ -364,6 +372,72 @@ public void testCanListenForQueryMetadataChanges() {
     listener2.remove();
   }
 
+  @Test
+  public void testInvalidQueryFilters() {
+    CollectionReference collection = testCollection();
+    // Multiple inequalities, one of which is inside a nested composite filter.
+    assertThrows(
+        IllegalArgumentException.class,
+        () -> {
+          Query query =
+              collection
+                  .where(
+                      or(
+                          and(equalTo("a", "b"), greaterThan("c", "d")),
+                          and(equalTo("e", "f"), equalTo("g", "h"))))
+                  .where(greaterThan("r", "s"));
+        });
+    // OrderBy and inequality on different fields. Inequality inside a nested composite filter.
+    assertThrows(
+        IllegalArgumentException.class,
+        () -> {
+          Query query =
+              collection
+                  .where(
+                      or(
+                          and(equalTo("a", "b"), greaterThan("c", "d")),
+                          and(equalTo("e", "f"), equalTo("g", "h"))))
+                  .orderBy("r");
+        });
+    // Conflicting operations within a composite filter.
+    assertThrows(
+        IllegalArgumentException.class,
+        () -> {
+          Query query =
+              collection.where(
+                  or(
+                      and(equalTo("a", "b"), inArray("c", Arrays.asList("d", "e"))),
+                      and(equalTo("e", "f"), notInArray("c", Arrays.asList("d", "e")))));
```

Comment:
Hm, I don't think we want to go down this level and provide semantic validation. This is just a huge can of worms - I think we should reject "syntactically" invalid queries but not inspect the actual contents of the query operations.

Here and below.

Reply:
I think I don't understand what you mean. We're not throwing the error because `in` and `not-in` have the same list values. We're throwing the error because `in` and `not-in` are conflicting operations.

if we see `or(a>1, b>1)`, we should reject it because it has multiple inequalities.
if we see `or(a in [1,2], b not-in [1,2])`, we should reject because of conflicting ops.
right?

Reply:
Okay - if the query cannot be run on the backend no mattter what values we are filtering by then yes, we should reject it. Sorry for the misunderstanding (which was mostly driven by the fact that you used the same field values in for `in` and `notIn`.

Reply:
yeah that's a bit confusing. I've changed the field values in the test